File: src/main/java/com/oasis/social/persistence/IUserPersistence.java
package com.oasis.social.persistence;

import com.oasis.social.models.User;
import io.vertx.core.Future;
import io.vertx.sqlclient.Row;
import io.vertx.sqlclient.RowSet;
import io.vertx.sqlclient.SqlResult;

import java.util.List;

public interface IUserPersistence {
  static IUserPersistence create() {
    return new UserPersistenceImpl();
  }

  Future<List<User>> findUsers();

  Future<User> findUserById(String userId);

  Future<RowSet<Row>> addUser(User user);

  void updateUser(User user);

  Future<SqlResult<Void>> deleteUserById(String userId);
}

File: src/main/kotlin/com/oasis/social/persistence/IUserPersistence.kt
package com.oasis.social.persistence

import com.oasis.social.models.User
import com.oasis.social.persistence.impl.UserPersistenceImpl
import io.vertx.core.Future

interface IUserPersistence {
  companion object {
    fun create(): IUserPersistence {
      return UserPersistenceImpl()
    }
  }

  fun findUsers(): Future<Collection<User>>

  fun findUserById(userId: String): User?

  fun addUser(user: User)

  fun updateUser(user: User)

  fun deleteUserById(userId: String)
}
